[Fix] Issue#2408

Cesar-Medeiros opened the pull request:
## Description

Enabling the user to open folders containing files, or files themselves, created by the application.

![Open file](https://user-images.githubusercontent.com/28062127/50244541-17d9c100-03c8-11e9-85aa-c5c456021691.gif)


## Issue fixed

[#2408](https://github.com/BoostIO/Boostnote/issues/2408)

## Type of changes

- :white_circle: Bug fix (Change that fixed an issue)
- :white_circle: Breaking change (Change that can cause existing functionality to change)
- :white_circle: Improvement (Change that improves the code. Maybe performance or development improvement)
- :radio_button: Feature (Change that adds new functionality)
- :white_circle: Documentation change (Change that modifies documentation. Maybe typo fixes)

## Checklist:

- :radio_button: My code follows [the project code style](docs/code_style.md)
- :radio_button: I have written test for my code and it has been tested
- :radio_button: All existing tests have been passed
- :radio_button: I have attached a screenshot/video to visualize my change if possible

Margarida-Silva commented:
We would really appreciate to know whether this PR is ready to be accepted, hence we are working on this open source project in the context of a university course and it would be beneficial to discuss this PR, or even see it being approved.

ZeroX-DG commented:
@Margarida-Silva I'll review and approve this PR soon for your uni course so if you can provide me with more information about this PR I'll be able to speed up the review process. I haven't gone to uni yet (I will next year) so I don't know how much time you can wait for this PR to be approved so please don't hesitate to tell me if you need this PR to be reviewed faster.

Cesar-Medeiros commented:
@ZeroX-DG Thanks for your reply. In this PR we added the functionality of opening a file instead of just importing it. Each time a file is opened by clicking on "Open File" we can change the note in boostnote and the original file is also modified. The same happens if we edit the original file, the note will be modified in boostnote. The gif attached in the main post helps to understand this PR.

Thanks for speeding up the review process.

ZeroX-DG commented on browser/main/NoteList/index.js:
```diff
@@ -128,12 +132,56 @@ class NoteList extends React.Component {
     const note = this.notes[0]
     const prevKey = prevProps.location.query.key
     const noteKey = visibleNoteKeys.includes(prevKey) ? prevKey : note && note.key
+    const CSON = require('@rokt33r/season')
```
Can you move this to the top of the file?

ZeroX-DG commented on browser/main/NoteList/index.js:
```diff
@@ -128,12 +132,56 @@ class NoteList extends React.Component {
+      if (focusNote.filepath !== undefined) {
+        try {
+          this.fsWatcher.close()
+        } catch (e) {
```
Please at least `console.error` the error. It might come in handy in the future.

ZeroX-DG commented on browser/main/NoteList/index.js:
```diff
@@ -128,12 +132,56 @@ class NoteList extends React.Component {
+
+            case 'rename':
+              updatedNote.filepath = undefined
+              this.fsWatcher.close()
```
Why did you have to catch for errors in the previous line 150 and not here?

ZeroX-DG commented:
Please fix the conflict and also I encounter this error when I run the app.
![screenshot from 2019-01-04 22-36-59](https://user-images.githubusercontent.com/12984316/50696558-56cb7000-1072-11e9-984c-79d194b3cba9.png)

ZeroX-DG commented:
I'm sorry @Cesar-Medeiros, this is the PR that causing the bug above: https://github.com/BoostIO/Boostnote/pull/2768, please merge the master branch again to fix this bug. Also, I found 2 bugs related to your PR.

**First bug**: When I add a new file, the highlight for that note is wrong. Demo:

![boostnote bug 1](https://user-images.githubusercontent.com/12984316/50731537-1404b780-119a-11e9-9943-e561c9f91bd2.gif)

**Second bug**: After I added that new file, I tried to update its' content using Boostnote and then using external editor. After that I switch to a new note and the app crashed and this error printed in the terminal (This is probably a Linux related bug):
```
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
/home/zerox/Desktop/Projects/Boostnote/node_modules/electron/dist/electron --type=renderer --no-sandb[8119]: ../../vendor/node/src/fs_event_wrap.cc:91:static void node::(anonymous namespace)::FSEventWrap::GetInitialized(const FunctionCallbackInfo<v8::Value> &): Assertion `wrap != nullptr' failed.
 1: node::Abort() [/home/zerox/Desktop/Projects/Boostnote/node_modules/electron/dist/libnode.so]
 2: 0x7f347fd4a422 [/home/zerox/Desktop/Projects/Boostnote/node_modules/electron/dist/libnode.so]
 3: 0x7f347fd3f5fa [/home/zerox/Desktop/Projects/Boostnote/node_modules/electron/dist/libnode.so]
 4: 0x295cd9c07841
```

Cesar-Medeiros commented:
I will try to fix this bugs as soon as possible. 
Thank you.

Cesar-Medeiros commented:
@ZeroX-DG I believe I have fixed both bugs.
We deeply appreciate your concern on reviewing this PR as soon as possible.
Regarding the university course context, we would also be pleased to hear from you about another issue we've attempted to fix (#2702).

ZeroX-DG commented:
@Cesar-Medeiros Editing in Boostnote update the file content works fine but now when I edit the note using an external app, it won't update in Boostnote, sometimes the app crashed and the second bug appears. Also, when I rename the file and switch to other note before switching back to that note, the app carshed and again the second bug appears.

ZeroX-DG commented:
The problem seems to occur only when I edit with Linux gedit app.

ZeroX-DG commented:
@Cesar-Medeiros do you still interested in this PR or should I close it?

Cesar-Medeiros commented:
@ZeroX-DG, Our team is no longer working on this issue. You can close it.
